File: real-estate-system/backend/src/main/java/com/realestate/api/entity/Contrat.java
package com.realestate.api.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "Contrat")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Contrat {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, updatable = false)
    private LocalDateTime dateCreation;

    @Column
    private LocalDateTime dateModification;

    @Enumerated(EnumType.STRING)
    @Column(length = 50, nullable = false)
    private StatutContrat statut = StatutContrat.EN_COURS;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "bien_id", nullable = false)
    private Bien bien;

    @Enumerated(EnumType.STRING)
    @Column(name = "type_contrat", length = 50, nullable = false)
    private TypeContrat typeContrat;

    // Snapshot fields — copied from offer at contract creation time
    @Column(name = "snap_mensualite", precision = 15, scale = 2)
    private BigDecimal snapMensualite;

    @Column(name = "snap_caution", precision = 15, scale = 2)
    private BigDecimal snapCaution;

    @Column(name = "snap_duree_mois")
    private Integer snapDureeMois;

    @Column(name = "snap_prix", precision = 15, scale = 2)
    private BigDecimal snapPrix;

    @Column(name = "snap_date_dispo")
    private LocalDate snapDateDispo;

    @Column(length = 500)
    private String documentSigne;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "compte_createur_id")
    private Compte createdBy;

    @OneToMany(mappedBy = "contrat", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Cosigner> cosigners = new ArrayList<>();

    @PrePersist
    protected void onCreate() {
        dateCreation = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        dateModification = LocalDateTime.now();
    }

    // Helper methods

    public void addCosigner(Cosigner cosigner) {
        cosigners.add(cosigner);
        cosigner.setContrat(this);
    }

    public void removeCosigner(Cosigner cosigner) {
        cosigners.remove(cosigner);
        cosigner.setContrat(null);
    }

    public boolean isRentalContract() {
        return typeContrat == TypeContrat.LOCATION;
    }

    public boolean isPurchaseContract() {
        return typeContrat == TypeContrat.ACHAT;
    }

    public TypeContrat getType() {
        return typeContrat;
    }

    // Enums

    public enum StatutContrat {
        EN_COURS,
        SIGNE,
        ANNULE,
        TERMINE
    }

    public enum TypeContrat {
        LOCATION,
        ACHAT
    }
}

File: real-estate-system/frontend-mobile/lib/models/contrat.dart
class Contrat {
  final int id;
  final String? dateCreation;
  final String? dateModification;
  final String statut;
  final String type;
  final int? bienId;
  final String? typeContrat;
  final Map<String, dynamic>? bien;
  final bool hasSignedDocument;
  final List<CosignerDTO> cosigners;

  // Snapshot fields (copied from offer at contract creation)
  final double? snapMensualite;
  final double? snapCaution;
  final int? snapDureeMois;
  final double? snapPrix;
  final String? snapDateDispo;

  Contrat({
    required this.id,
    this.dateCreation,
    this.dateModification,
    this.statut = '',
    this.type = '',
    this.bienId,
    this.typeContrat,
    this.bien,
    this.hasSignedDocument = false,
    this.cosigners = const [],
    this.snapMensualite,
    this.snapCaution,
    this.snapDureeMois,
    this.snapPrix,
    this.snapDateDispo,
  });

  factory Contrat.fromJson(Map<String, dynamic> json) {
    return Contrat(
      id: json['id'] as int,
      dateCreation: json['dateCreation'] as String?,
      dateModification: json['dateModification'] as String?,
      statut: json['statut'] as String? ?? '',
      type: json['type'] as String? ?? '',
      bienId: json['bienId'] as int? ?? (json['bien'] as Map<String, dynamic>?)?['id'] as int?,
      typeContrat: json['typeContrat'] as String? ?? json['type'] as String?,
      bien: json['bien'] as Map<String, dynamic>?,
      hasSignedDocument: json['hasSignedDocument'] as bool? ?? false,
      cosigners: (json['cosigners'] as List?)
              ?.map((c) => CosignerDTO.fromJson(c as Map<String, dynamic>))
              .toList() ??
          [],
      snapMensualite: (json['snapMensualite'] as num?)?.toDouble(),
      snapCaution: (json['snapCaution'] as num?)?.toDouble(),
      snapDureeMois: json['snapDureeMois'] as int?,
      snapPrix: (json['snapPrix'] as num?)?.toDouble(),
      snapDateDispo: json['snapDateDispo'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      if (dateCreation != null) 'dateCreation': dateCreation,
      if (dateModification != null) 'dateModification': dateModification,
      'statut': statut,
      'type': type,
      if (bienId != null) 'bienId': bienId,
      if (typeContrat != null) 'typeContrat': typeContrat,
      if (bien != null) 'bien': bien,
      'hasSignedDocument': hasSignedDocument,
      'cosigners': cosigners.map((c) => c.toJson()).toList(),
      if (snapMensualite != null) 'snapMensualite': snapMensualite,
      if (snapCaution != null) 'snapCaution': snapCaution,
      if (snapDureeMois != null) 'snapDureeMois': snapDureeMois,
      if (snapPrix != null) 'snapPrix': snapPrix,
      if (snapDateDispo != null) 'snapDateDispo': snapDateDispo,
    };
  }
}

class CosignerDTO {
  final int? personneId;
  final String? nom;
  final String? prenom;
  final String typeSignataire;

  CosignerDTO({
    this.personneId,
    this.nom,
    this.prenom,
    required this.typeSignataire,
  });

  factory CosignerDTO.fromJson(Map<String, dynamic> json) {
    return CosignerDTO(
      personneId: json['personneId'] as int?,
      nom: json['nom'] as String?,
      prenom: json['prenom'] as String?,
      typeSignataire: json['typeSignataire'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (personneId != null) 'personneId': personneId,
      if (nom != null) 'nom': nom,
      if (prenom != null) 'prenom': prenom,
      'typeSignataire': typeSignataire,
    };
  }
}
